JAVA-3083 Doc updates for new features in 4.17.0 (#1677)
apache/cassandra-java-driver: upgrade_guide/README.md

## Upgrade guide

### 4.17.0

#### Java17 is now a supported platform

With [JAVA-3042](https://datastax-oss.atlassian.net/browse/JAVA-3042) the driver is now tested against Java17 as
well as the other LTS JVM releases Java8 and Java11.  Based on our testing the Java driver appears to function
normally on Java17, but of course bugs are always a possibility.  Issues arising from the Java17 platform will be

> triaged and addressed just like any other issue.

Author: I'm not super-happy with the way this language came out.  If anybody has any suggestions on a way to say all this in something that sounds more like English I'm all ears.

Reviewer: I'm happy with this from a technical sentiment pov

> If you discover an issue with the Java driver running on Java 17, please let us know. We will triage and address Java 17 issues.

Reviewer 2: How are they supposed to let us know?

Author: Mentioned above, reposting here:

"Any of the regular reporting mechanisms would work. Probably the default way would be to contact support but posting tickets to JIRA, posting to the Google group/mailing list, etc.... any of those would be fine.

I didn't specify 'cause I didn't want to create the perception of limiting it to any one mechanism. I'm open to other ideas here, though."

#### Updated API for vector search

> The 4.16.0 release introduced support for the CQL vector datatype (see

Reviewer: Nit: do we mention anywhere which cassandra/dse versions are needed for CqlVector? If not should we? I realized when copy and pasting your example into an integration test that I don't know what version of cassandra I need

Author: Eventually yes, but for now the code isn't released in any Cassandra or DSE version.  It will be available in Astra but nowhere else yet.

So eventually yeah, we'll prolly need to add more detailed information regarding dependencies... but I don't think we're there yet.

The 4.16.0 release introduced support for the CQL `vector` datatype. This release modifies the `CqlVector`
value type used to represent a CQL vector to make it easier to use.  `CqlVector` now implements the Iterable interface
as well as several methods modelled on the JDK's List interface. For more, see
[JAVA-3060](https://datastax-oss.atlassian.net/browse/JAVA-3060). 

The builder interface was replaced with factory methods that resemble similar methods on `CqlDuration`.
For example, the following code will create a keyspace and table, populate that table with some data, and then execute

> a query which will return a `vector` type.  This data is retrieved directly via `Row.getVector()` and the resulting

Reviewer 2: ```suggestion
a query that will return a `vector` type.  This data is retrieved directly via `Row.getVector()` and the resulting
```

`CqlVector` value object can be interrogated directly.

```java
try (CqlSession session = new CqlSessionBuilder().withLocalDatacenter("datacenter1").build()) {

>     session.execute("drop keyspace if exists test");

Reviewer: nit: is the drop keyspace line neccessary for this example?

Author: Perhaps not strictly necessary but I'd argue it's certainly a good practice.   Including it here communicates the idea that this example should reset its state before it runs; I want to cut off any cases of the test code not working because the user already had a "test" keyspace defined.

    session.execute("create table test.foo(i int primary key, j vector<float, 3>)");
    session.execute("create custom index ann_index on test.foo(j) using 'StorageAttachedIndex'");
    session.execute("insert into test.foo (i, j) values (1, [8, 2.3, 58])");
    session.execute("insert into test.foo (i, j) values (2, [1.2, 3.4, 5.6])");
    session.execute("insert into test.foo (i, j) values (5, [23, 18, 3.9])");

>     ResultSet rs=session.execute("select j from test.foo where j ann of [3.4, 7.8, 9.1] limit 1");

Reviewer: Nit: upper case CQL keywords to match the style used elsewhere in this file

Author: Yeah, that's reasonable.

Updated.

    for(Row row:rs){
        CqlVector<Float> v = row.getVector(0,Float.class);
        System.out.println(v);
        if(Iterables.size(v) != 3){
            throw new RuntimeException("Expected vector with three dimensions");
        }
    }

> }

Reviewer: Nit: Fix spacing in this code to conform to our usual code standards

Author: The samples in this doc appear to vary quite a bit.  Most use an indent of 4 spaces (which is what I use here) while a few spots use an indent of two spaces.  I find a 4 space indent to be more readable as well so I think I'll update all the samples in the doc to use a consistent 4 space indent.

Reviewer: To clarify I was mostly thinking about the spacing within the lines, eg between variable and `=`, in the loop syntax etc

Author: Makes sense @hhughes.  It's probably a good exercise to standardize the docs on a (mostly) uniform Java code style anyway. :)

## Upgrade guide

### 4.17.0

#### Beta support for Java17

> With the completion of [JAVA-3042](https://datastax-oss.atlassian.net/browse/JAVA-3042) the driver now passes our
> test suite when run on Java17.  We also now include Java17 in our automated test matrix for Java driver releases.
> All features appear to function normally when run with Java17 but we do not want to declare full support for this
> platform until we've received feedback from other users in the ecosystem.
> 
> If you discover an issue with the Java driver running on Java17 please let us know!  Java17 issues will be triaged
> and addressed just like issues from any supported Java version.

Reviewer 2: ```suggestion
With the completion of [JAVA-3042](https://datastax-oss.atlassian.net/browse/JAVA-3042), the driver now passes our automated test matrix for Java driver releases.
While all features function normally when run with Java 17 tests, we do not offer full support for this
platform until we've received feedback from other users in the ecosystem.

If you discover an issue with the Java driver running on Java 17, please let us know. We will traige and address Java 17 issues.
```

Reviewer 2: How should they let us know? Is there an email address we can include?

Author: Any of the regular reporting mechanisms would work.  Probably the default way would be to contact support but posting tickets to JIRA, posting to the Google group/mailing list, etc.... any of those would be fine.

I didn't specify 'cause I didn't want to create the perception of limiting it to any one mechanism.  I'm open to other ideas here, though.

#### Updated API for vector search

The 4.16.0 release introduced support for the CQL vector datatype (see
[JAVA-3060](https://datastax-oss.atlassian.net/browse/JAVA-3060) for more detail).  This release modifies the CqlVector
value type used to represent a CQL vector to make it easier to use.  CqlVector now implements the Iterable interface
as well as several methods modelled on the JDK's List interface, so it should have a familiar feel to it.  We've also
done away with the builder interface, replacing it with some factory methods that resemble similar methods on CqlDuration.

As an example, the following code will create a keyspace and table, populate that table with some data and then execute
a query which will return a vector type.  This data is retrieved directly via Row.getVector() and the resulting

> CqlVector value object can be interrogated directly.

Reviewer 2: ```suggestion
The 4.16.0 release introduced support for the CQL `vector` datatype. This release modifies the `CqlVector`
value type used to represent a CQL vector to make it easier to use.  `CqlVector` now implements the Iterable interface
as well as several methods modelled on the JDK's List interface. For more, see
[JAVA-3060](https://datastax-oss.atlassian.net/browse/JAVA-3060). 

The builder interface was replaced with factory methods that resemble similar methods on `CqlDuration`.
For example, the following code will create a keyspace and table, populate that table with some data, and then execute
a query which will return a `vector` type.  This data is retrieved directly via `Row.getVector()` and the resulting
`CqlVector` value object can be interrogated directly.
```

```java
try (CqlSession session = new CqlSessionBuilder().withLocalDatacenter("datacenter1").build()){

    session.execute("drop keyspace if exists test");
    session.execute("CREATE KEYSPACE test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}");
    session.execute("CREATE TABLE test.foo(i int primary key, j vector<float, 3>)");
    session.execute("CREAT CUSTOM INDEX ann_index ON test.foo(j) USING 'StorageAttachedIndex'");
    session.execute("INSERT INTO test.foo (i, j) VALUES (1, [8, 2.3, 58])");
    session.execute("INSERT INTO test.foo (i, j) VALUES (2, [1.2, 3.4, 5.6])");
    session.execute("INSERT INTO test.foo (i, j) VALUES (5, [23, 18, 3.9])");
    ResultSet rs=session.execute("SELECT j FROM test.foo WHERE j ann of [3.4, 7.8, 9.1] limit 1");

>     for (Row row:rs){

Reviewer: Nit: spacing around `:` and before  `{` (and also on lines 28 + 41)

Author: Agreed, will fix.  Something in one of the automated formatting processes did something to the sample code in this readme... I think it might've been IntelliJ actually.  Everything got heavily concatenated. :(

>         CqlVector<Float> v = row.getVector(0,Float.class);

Reviewer: Nit: spacing after `,`

        System.out.println(v);

>         if (Iterables.size(v) != 3){

Reviewer: Nit: Is this size check necessary for the code snippet?

Author: It's not strictly necessary.  I added it here to show that v implements Iterable and can be used in anything expecting that interface.  The other obvious way to do that was to iterate over it in a for loop but it wasn't clear to me that that would be a useful exercise for a vector.

            throw new RuntimeException("Expected vector with three dimensions");
        }
    }
}
```

> You can also use the CqlVector type with prepared statements:

Reviewer 2: ```suggestion
You can also use the `CqlVector` type with prepared statements:
```

```java

> PreparedStatement preparedInsert = session.prepare("insert into test.foo (i, j) values (?,?)");

Reviewer: Nit: casing of cql keywords

CqlVector<Float> vector = CqlVector.newInstance(1.4f, 2.5f, 3.6f);
session.execute(preparedInsert.bind(3, vector));
```

In some cases it may be more appropriate to access the vector directly as an array of some numerical type.  This version
supports such use cases by providing a codec which translates a CQL vector to and from a primitive array.  In this release
only float arrays are supported but additional codecs may be added for other primitive numerical types in future releases.

> You can find more information about this codec in the manual documentation on [custom codecs](../manual/core/custom_codecs/)

Reviewer 2: ```suggestion
In some cases, it makes sense to access the vector directly as an array of some numerical type. This version
supports such use cases by providing a codec which translates a CQL vector to and from a primitive array. Only float arrays are supported. 
You can find more information about this codec in the manual documentation on [custom codecs](../manual/core/custom_codecs/)
```
